src/openrct2/world/sprite.c

```diff
@@ -823,3 +826,106 @@ bool sprite_get_flashing(rct_sprite *sprite)
     assert(sprite->unknown.sprite_index < MAX_SPRITES);
     return _spriteFlashingList[sprite->unknown.sprite_index];
 }
+
+static bool sprite_is_in_cycle(uint16 sprite_idx, bool fix)
+{
+    if (sprite_idx == SPRITE_INDEX_NULL)
+    {
+        return false;
+    }
+    const rct_sprite * fast = get_sprite(sprite_idx);
+    const rct_sprite * slow = fast;
+    bool increment_slow = false;
+    bool cycled = false;
+    while (fast->unknown.sprite_index != SPRITE_INDEX_NULL)
+    {
+        // increment fast every time, unless reached the end
+        if (fast->unknown.next == SPRITE_INDEX_NULL)
+        {
+            break;
+        }
+        else {
+            fast = get_sprite(fast->unknown.next);
+        }
+        // increment slow only every second iteration
+        if (increment_slow)
+        {
+            slow = get_sprite(slow->unknown.next);
+        }
+        increment_slow = !increment_slow;
+        if (fast == slow)
+        {
+            cycled = true;
+            if (fix)
+            {
+                rct_sprite * spr = get_sprite(sprite_idx);
+                // There may be a better solution than simply setting this to 0xFFFF
+                spr->unknown.next = SPRITE_INDEX_NULL;
+            }
+            break;
+        }
+    }
+    return cycled;
+}
+
+static bool sprite_is_in_cycle_quadrant(uint16 sprite_idx, bool fix)
+{
+    if (sprite_idx == SPRITE_INDEX_NULL)
+    {
+        return false;
+    }
+    const rct_sprite * fast = get_sprite(sprite_idx);
+    const rct_sprite * slow = fast;
+    bool increment_slow = false;
+    bool cycled = false;
+    while (fast->unknown.sprite_index != SPRITE_INDEX_NULL)
+    {
+        // increment fast every time, unless reached the end
+        if (fast->unknown.next_in_quadrant == SPRITE_INDEX_NULL)
+        {
+            break;
+        }
+        else {
+            fast = get_sprite(fast->unknown.next_in_quadrant);
+        }
+        // increment slow only every second iteration
+        if (increment_slow)
+        {
+            slow = get_sprite(slow->unknown.next_in_quadrant);
+        }
+        increment_slow = !increment_slow;
+        if (fast == slow)
+        {
+            cycled = true;
+            if (fix)
+            {
+                rct_sprite * spr = get_sprite(sprite_idx);
+
+                // There may be a better solution than simply setting this to 0xFFFF
+                spr->unknown.next_in_quadrant = SPRITE_INDEX_NULL;
```

Comment:
I did look for a better solution here, and also in `sprite_is_in_cycle()` but was unable to find something that would work (most functions rely on the sprite being 'sane'). I was tempted to adapt one (e.g. using a wrapper), but thought I'd seek input first.